packages/astro/src/content/content-layer.ts

```diff
@@ -267,44 +267,121 @@ class ContentLayer {
 					return;
 				}
 
-				const collectionWithResolvedSchema = { ...collection, schema };
-
-				const parseData: LoaderContext['parseData'] = async ({ id, data, filePath = '' }) => {
-					const { data: parsedData } = await getEntryDataAndImages(
-						{
-							id,
-							collection: name,
-							unvalidatedData: data,
-							_internal: {
-								rawData: undefined,
-								filePath,
-							},
-						},
-						collectionWithResolvedSchema,
-						false,
-					);
-
-					return parsedData;
-				};
-
 				const context = await this.#getLoaderContext({
 					collectionName: name,
-					parseData,
 					loaderName: collection.loader.name,
 					refreshContextData: options?.context,
 				});
 
+				const realStore = this.#store.scopedStore(name);
+
 				if (typeof collection.loader === 'function') {
-					return simpleLoader(collection.loader as CollectionLoader<{ id: string }>, context);
+					await simpleLoader(collection.loader as CollectionLoader<{ id: string }>, {
+						...context,
+						store: realStore,
+					});
+					return;
 				}
 
 				if (!collection.loader.load) {
 					throw new Error(`Collection loader for ${name} does not have a load method`);
 				}
 
-				return collection.loader.load(context);
+				const memoryStore = new InMemoryDataStore(new Map(realStore.entries()));
+
+				const result = collection.loader.load({
+					...context,
+					store: memoryStore,
+				});
+
+				if (!schema && result?.schema) {
+					schema = result.schema;
+				}
+
+				return {
+					schema,
+					types: result?.types,
+					realStore,
+					memoryStore,
+					name,
+					collection,
+				};
 			}),
 		);
+
+		const filteredRawLoaderResults = rawLoaderResults.filter((result) => !!result);
+
+		await Promise.all(
+			filteredRawLoaderResults.map(async ({ memoryStore, realStore, schema, name, collection }) => {
+				for (const event of memoryStore.events) {
+					switch (event.type) {
+						case 'set': {
+							let data = event.entry.data;
+							if (schema) {
+								data = await getEntryData(
+									{
+										id: event.entry.id,
+										collection: name,
+										unvalidatedData: data,
+										_internal: {
+											rawData: undefined,
+											filePath: event.entry.filePath!,
+										},
+									},
+									{
+										...collection,
+										schema,
+									},
+									false,
+								);
+
+								new Traverse(data).forEach((ctx, value) => {
+									if (!isReference(value)) {
+										return;
+									}
+									// TODO: better data structure
+									const collectionStore = filteredRawLoaderResults.find(
+										(e) => e.name === value.collection,
+									)!.memoryStore;
+									const id = 'id' in value ? value.id : value.slug;
+									if (!collectionStore.has(id)) {
+										// TODO: AstroError
+										throw new Error(
+											`Invalid reference for ${name}.${ctx.keys!.map((key) => key.toString()).join('.')}: cannot find entry ${id}`,
+										);
+									}
+								});
```

Comment:
Is there no way to do this inside the schema validation? Traversing always seems a bit of a hack

Reply:
Yeah I agree. If reference was passed like `image()` we could have some context in there but AFAIK with the current implementation there's not much we can do